tiered-storage/jcloud/src/main/java/org/apache/bookkeeper/mledger/offload/jcloud/provider/JCloudBlobStoreProvider.java

```diff
@@ -0,0 +1,233 @@
+            return BLOB_STORE_BUILDER.getBlobStore(config);
+        }
+
+        @Override
+        public void buildCredentials(TieredStorageConfiguration config) {
+            if (config.getProviderCredentials() == null) {
+                try {
+                    String gcsKeyContent = Files.toString(
+                            new File(config.getConfigProperty("gcsManagedLedgerOffloadServiceAccountKeyFile")),
+                                     Charset.defaultCharset());
+                    config.setProviderCredentials(new GoogleCredentialsFromJson(gcsKeyContent).get());
+                } catch (IOException ioe) {
+                    log.error("Cannot read GCS service account credentials file: {}",
+                            config.getConfigProperty("gcsManagedLedgerOffloadServiceAccountKeyFile"));
+                    throw new IllegalArgumentException(ioe);
+                }
+            }
+        }
+    },
+
+    TRANSIENT("transient", new AnonymousProviderMetadata(new TransientApiMetadata(), "")) {
+        @Override
+        public void validate(TieredStorageConfiguration config) throws IllegalArgumentException {
+            if (Strings.isNullOrEmpty(config.getBucket())) {
+                throw new IllegalArgumentException(
+                    "Bucket cannot be empty for Local offload");
+            }
+        }
+
+        @Override
+        public BlobStore getBlobStore(TieredStorageConfiguration config) {
+            if (!BlobStoreRepository.containsKey(config.getBlobStoreLocation())) {
+                BlobStore bs = ContextBuilder.newBuilder(config.getDriver().toLowerCase())
+                        .buildView(BlobStoreContext.class)
+                        .getBlobStore();
+
+                if (!bs.containerExists(config.getBucket())) {
+                    Location loc = new LocationBuilder()
+                            .scope(LocationScope.HOST)
+                            .id(UUID.randomUUID() + "")
+                            .description("Transient " + config.getBucket())
+                            .build();
+
+                    bs.createContainerInLocation(loc, config.getBucket());
+                }
+                BlobStoreRepository.put(config.getBlobStoreLocation(), bs);
+            }
+            return BlobStoreRepository.get(config.getBlobStoreLocation());
+        }
+
+        @Override
+        public void buildCredentials(TieredStorageConfiguration config) {
+            // No-op
+        }
+    };
+
+    public static final boolean driverSupported(String driverName) {
+        for (JCloudBlobStoreProvider provider: JCloudBlobStoreProvider.values()) {
+            if (provider.getDriver().equalsIgnoreCase(driverName)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private String driver;
+    private ProviderMetadata providerMetadata;
+
+    JCloudBlobStoreProvider(String s, ProviderMetadata providerMetadata) {
```

Comment:
From what I see, this is more configuration than metadata. this is the bucket and region, no?

Reply:
The driver field, is what is used to map from the value of the "managedLedgerOffloadDriver" property in the broker.conf file to one of these Enumerations.

The ProviderMetadata field, is an instance of [org.jclouds.providers.ProviderMetadata](url
https://jclouds.apache.org/reference/javadoc/2.1.x/org/jclouds/providers/ProviderMetadata.html), which has different concrete implementations based on the JClouds provider, e.g AWSS3ProviderMetadata, GoogleCloudStorageProviderMetadata, or AnonymousProviderMetadata.  

These instances also support additional properties and configuration settings beyond the driver name.  The intent is to use these provider specific instances when we create the BlobStores via the `ContextBuilder.newBuilder(config.getProviderMetadata())` call